std/experimental/allocator/package.d

```diff
@@ -1123,48 +1123,107 @@ allocated with the same allocator.
 */
 void dispose(A, T)(auto ref A alloc, T* p)
 {
+    if (!p) return;
     static if (hasElaborateDestructor!T)
-    {
-        destroy(*p);
-    }
+        p.__dtor;
     alloc.deallocate((cast(void*)p)[0 .. T.sizeof]);
 }
 
 /// Ditto
 void dispose(A, T)(auto ref A alloc, T p)
-if (is(T == class) || is(T == interface))
+if (is(T == class) && T.stringof != Object.stringof)
+{
+    if (!p) return;
+    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
+    static if (__traits(hasMember, T, "__dtor"))
```

Comment:
It's possible to detect the case and then to dispatch to the "dynamic class disposer", however I have to investiguate more to see if the derived type can be found at CT. It would be a pity not to allow `@nogc` destructor for this typical case.